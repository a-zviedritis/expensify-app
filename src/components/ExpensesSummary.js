import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ hiddenExpenseCount, visibleExpenseCount, visibleExpensesTotal }) => {
  const hiddenExpenseWord = hiddenExpenseCount === 1 ? 'expense': 'expenses';
  const visibleExpenseWord = visibleExpenseCount === 1 ? 'expense': 'expenses';
  const formattedExpensesTotal = numeral(visibleExpensesTotal / 100).format('0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{visibleExpenseCount}</span> {visibleExpenseWord} totalling <span>â‚¬{formattedExpensesTotal}</span></h1>
        {hiddenExpenseCount > 0 && <p className="page-header__sub-title">{hiddenExpenseCount} {hiddenExpenseWord} filtered out</p>}
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    hiddenExpenseCount: state.expenses.length - visibleExpenses.length,
    visibleExpenseCount: visibleExpenses.length,
    visibleExpensesTotal: selectExpensesTotal(visibleExpenses),
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
