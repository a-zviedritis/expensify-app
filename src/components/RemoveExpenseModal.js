import React from 'react';
import Modal from 'react-modal';

const RemoveExpenseModal = (props) => (
  <Modal 
    isOpen={!!props.expenseForRemoval}
    onRequestClose={props.handleDenyRemoval}
    contentLabel="Remove expense"
    closeTimeoutMS={50}
    className="modal"
  >
    <h3 className="modal__title">Remove expense?</h3>
    <div className="modal__selection">
      <button className="button" onClick={props.handleConfirmRemoval}>Yes</button>
      <button className="button" onClick={props.handleDenyRemoval}>No</button>
    </div>
  </Modal>
);

export default RemoveExpenseModal;