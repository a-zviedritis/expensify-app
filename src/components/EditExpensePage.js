import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import RemoveExpenseModal from './RemoveExpenseModal';

export class EditExpensePage extends React.Component {
  state = {
    expenseForRemoval: undefined
  }
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  markForRemoval = () => {
    this.setState(() => ({
      expenseForRemoval: this.props.expense
    }));
  };
  handleConfirmRemoval = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  handleDenyRemoval = () => {
    this.setState(() => ({
      expenseForRemoval: undefined,
    }));
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button 
            className="button button--secondary"
            onClick={this.markForRemoval}
          >
            Remove Expense
          </button>
          <RemoveExpenseModal 
            expenseForRemoval={this.state.expenseForRemoval}
            handleConfirmRemoval={this.handleConfirmRemoval}
            handleDenyRemoval={this.handleDenyRemoval}
            history={this.props.history}
          />
        </div>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);