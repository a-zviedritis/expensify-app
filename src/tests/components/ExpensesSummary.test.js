import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[ expenses[0] ]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly when no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});