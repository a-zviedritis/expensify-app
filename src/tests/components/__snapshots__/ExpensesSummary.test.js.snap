// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     expense
     totalling â‚¬
    1.95
  </p>
</div>
`;

exports[`should render ExpensesSummary correctly when no expenses 1`] = `
<div>
  <p>
    No expenses to view
  </p>
</div>
`;
